python/kserve/test/test_grpc_server.py

```diff
@@ -0,0 +1,662 @@
+                "GRPCInferenceService"
+            ].methods_by_name["ModelInfer"]
+        ),
+        invocation_metadata={},
+        request=request,
+        timeout=20,
+    )
+
+    response, _, code, _ = model_infer_method.termination()
+    response = await response
+    response_dict = MessageToDict(
+        response,
+        preserving_proto_field_name=True,
+    )
+    assert code == grpc.StatusCode.OK
+    assert response_dict == {
+        "model_name": "FP16OutputModel",
+        "id": "123",
+        "outputs": [
+            {
+                "name": "fp16_output",
+                "datatype": "FP16",
+                "shape": ["8"],
+                "parameters": {"binary_data_size": {"int64_param": "16"}},
+            },
+            {
+                "name": "fp32_output",
+                "datatype": "FP32",
+                "shape": ["8"],
+                "parameters": {"binary_data_size": {"int64_param": "32"}},
+            },
+        ],
+        "raw_output_contents": [
+            "zUaaQc1Emj0ARs1CgERmPg==",
+            "mpnZQDMzM0CamZlAMzOzPwAAwECamVlAAACQQM3MzD8=",
+        ],
+    }
+
+
+@pytest.mark.asyncio
+@patch(
+    "kserve.protocol.grpc.servicer.to_headers", return_value=[]
+)  # To avoid NotImplementedError from trailing_metadata function
+async def test_grpc_fp16_input(mock_to_headers, server):
+    raw_input_contents = [
+        np.array([6.8, 2.8, 4.8, 1.4, 6.0, 3.4, 4.5, 1.6], dtype=np.float16).tobytes()
+    ]
+    request = grpc_predict_v2_pb2.ModelInferRequest(
+        model_name="FP16InputModel",
+        id="123",
+        inputs=[
+            {
+                "name": "fp16_input",
+                "shape": [2, 4],
+                "datatype": "FP16",
+            },
+        ],
+        raw_input_contents=raw_input_contents,
+    )
+
+    model_infer_method = server.invoke_unary_unary(
+        method_descriptor=(
+            grpc_predict_v2_pb2.DESCRIPTOR.services_by_name[
+                "GRPCInferenceService"
+            ].methods_by_name["ModelInfer"]
+        ),
+        invocation_metadata={},
+        request=request,
+        timeout=20,
+    )
+
+    response, _, code, _ = model_infer_method.termination()
+    response = await response
+    response_dict = MessageToDict(
+        response,
+        preserving_proto_field_name=True,
+    )
+    assert code == grpc.StatusCode.OK
+    assert response_dict == {
+        "model_name": "FP16InputModel",
+        "id": "123",
+        "outputs": [
+            {
+                "name": "int32_output",
+                "datatype": "INT64",
+                "shape": ["8"],
+                "parameters": {"binary_data_size": {"int64_param": "64"}},
+            },
+            {
+                "name": "fp16_output",
+                "datatype": "FP16",
+                "shape": ["8"],
+                "parameters": {"binary_data_size": {"int64_param": "16"}},
+            },
+        ],
+        "raw_output_contents": [
```

Comment:
Can we add the asserts to decode raw output content to InferResponse/InferOutput so deserializeBytesTensor is tested?

Reply:
Added